lib/segment/src/index/field_index/full_text_index/posting_list.rs

```diff
@@ -0,0 +1,162 @@
+use bitvec::prelude::*;
+
+use crate::types::PointOffsetType;
+
+#[derive(Clone, Debug)]
+pub enum PostingList {
+    FullVec(Vec<PointOffsetType>),
+    BitVec(BitVec),
+}
+
+impl Default for PostingList {
+    fn default() -> Self {
+        Self::FullVec(vec![])
+    }
+}
+
+impl PostingList {
+    const LOWER_THRESHOLD: f64 = 0.03;
+    const UPPER_THRESHOLD: f64 = 0.032;
+
+    pub fn new(idx: PointOffsetType) -> Self {
+        Self::FullVec(vec![idx])
+    }
+
+    pub fn insert(&mut self, idx: PointOffsetType, doc_count: usize) {
+        match self {
+            Self::FullVec(vec) => {
+                if let Err(insertion_idx) = vec.binary_search(&idx) {
+                    vec.insert(insertion_idx, idx);
```

Comment:
this is `O(N)`, do you assume that the vector is small enough, so we would do this?

Reply:
If my reasoning for the thresholds is true, it should be relatively small but only benchmarks can tell if the benefit of binary search outweighs the overhead of arbitrary insertion.

Reply:
Do you have an estimation of this sophisticated posting list VS BTreeSet?

Reply:
I haven't really tried that comparison. I'll check it out now and see how it goes.

Reply:
The results have come in. 
In terms of numbers: the BTreeSet implementation sits at 237MB while the current implementation achieves 183MB.
Here's the performance plot of the BTreeSet implementation:
![btreeset](https://user-images.githubusercontent.com/30220322/232638357-cfb0cd83-df4b-4b2a-b41d-272c67f8726e.png)
compared to the current implementation:
![bitvec](https://user-images.githubusercontent.com/30220322/232638376-bee1bc18-946a-40df-8eff-701125453de0.png)

Reply:
I tried benchmarking the current sorted document implementation and using an unsorted list and it seems that the sorted implementation outperforms the unsorted one each time by around 4-10%